pkg/spec/resources_transform.go

```diff
@@ -0,0 +1,275 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   v.Name,
+				Labels: app.Labels,
+			},
+			// since we updated the pvc spec before so this can be directly fed
+			// without having to do any addition extra
+			Spec: api_v1.PersistentVolumeClaimSpec(v.PersistentVolumeClaimSpec),
+		})
+	}
+	return pvcs, nil
+}
+
+func (app *ControllerFields) createSecrets() ([]runtime.Object, error) {
+	var secrets []runtime.Object
+
+	for _, s := range app.Secrets {
+		secret := &api_v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   s.Name,
+				Labels: app.Labels,
+			},
+			Data:       s.Data,
+			StringData: s.StringData,
+			Type:       s.Type,
+		}
+		secrets = append(secrets, secret)
+	}
+	return secrets, nil
+}
+
+// CreateK8sObjects, if given object DeploymentSpecMod, this function reads
+// them and returns kubernetes objects as list of runtime.Object
+// If the deployment is using field 'extraResources' then it will
+// also return file names mentioned there as list of string
+func (app *ControllerFields) CreateK8sObjects() ([]runtime.Object, []string, error) {
+	var objects []runtime.Object
+
+	if app.Labels == nil {
+		app.Labels = app.getLabels()
+	}
+
+	svcs, err := app.createServices()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Unable to create Kubernetes Service")
+	}
+
+	ings, err := app.createIngresses()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Unable to create Kubernetes Ingresses")
+	}
+
+	secs, err := app.createSecrets()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Unable to create Kubernetes Secrets")
+	}
+
+	app.PodSpec.Containers, err = populateContainers(app.Containers, app.ConfigMaps, app.Secrets)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "deployment %q", app.Name)
+	}
+	log.Debugf("object after population: %#v\n", app)
+
+	app.PodSpec.InitContainers, err = populateContainers(app.InitContainers, app.ConfigMaps, app.Secrets)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "deployment %q", app.Name)
+	}
+	log.Debugf("object after population: %#v\n", app)
+
+	// create pvc for each root level persistent volume
+	pvcs, err := app.createPVC()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to create Persistent Volume Claims")
+	}
+
+	vols, err := populateVolumes(app.PodSpec.Containers, app.VolumeClaims, app.PodSpec.Volumes)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "deployment %q", app.Name)
+	}
+	app.PodSpec.Volumes = append(app.PodSpec.Volumes, vols...)
+
+	var configMap []runtime.Object
+	for _, cd := range app.ConfigMaps {
+		cm := &api_v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   cd.Name,
+				Labels: app.Labels,
+			},
+			Data: cd.Data,
+		}
+
+		configMap = append(configMap, cm)
+	}
+
+	// please keep the order of the artifacts addition as it is
+
+	// adding non-controller objects
+	objects = append(objects, pvcs...)
+	log.Debugf("deployment: %s, pvc: %s\n", app.Name, spew.Sprint(pvcs))
```

Comment:
Again this should not be tied to `deployment` right? Because it is method tied to `ControllerFields` ? 

Also same comment applies to the rest of this method!

Reply:
yep, changed deployment to app for the subsequent lines as well.